pyuvdata/uvdata/ms.py

```diff
@@ -126,11 +886,715 @@ def find_not_ascii(s):
+        # handle this. For now, this just does a reverse lookup on the limited frames
+        # supported by UVData objects, although eventually it can be expanded to support
+        # more native MS frames.
+        reverse_dict = {ref: key for key, ref in COORD_UVDATA2CASA_DICT.items()}
+
+        try:
+            ref_name = reverse_dict[(str(frame_name), float(epoch_val))]
+        except KeyError:
+            raise ValueError(
+                "Frame %s (epoch %g) does not have a corresponding match to supported "
+                "frames in the measurement set file format" % (frame_name, epoch_val)
+            )
+
+        return ref_name
+
+    def _read_ms_main(
+        self, filepath, data_column, data_desc_dict, read_weights=True, flip_conj=False,
+    ):
+        """
+        Read data from the main table of a MS file.
+
+        This method is not meant to be called by users, and is instead a utility
+        function for the `read_ms` method (which users should call instead).
+
+        Parameters
+        ----------
+        filepath : str
+            The measurement set root directory to read from.
+        data_column : str
+            name of CASA data column to read into data_array. Options are:
+            'DATA', 'MODEL', or 'CORRECTED_DATA'
+        data_desc_dict : dict
+            Dictionary describing the various rows in the DATA_DESCRIPTION table of
+            an MS file. Keys match to the individual rows, and the values are themselves
+            dicts containing several keys (including "CORR_TYPE", "SPW_ID", "NUM_CORR",
+            "NUM_CHAN").
+        read_weights : bool
+            Read in the weights from the MS file, default is True. If false, the method
+            will set the `nsamples_array` to the same uniform value (namely 1.0).
+        flip_conj : bool
+            On read, whether to flip the conjugation of the baselines. Normally the MS
+            format is the same as what is used for pyuvdata (ant2 - ant1), hence the
+            default is False.
+
+        Returns
+        -------
+        spw_list : list of int
+            List of SPW numbers present in the data set, equivalent to the attribute
+            `spw_array` in a UVData object.
+        field_list : list of int
+            List of field IDs present in the data set. Matched to rows in the FIELD
+            table for the measurement set.
+        pol_list : list of int
+            List of polarization IDs (in the AIPS convention) present in the data set.
+            Equivalent to the attribute `polarization_array` in a UVData object.
+
+        Raises
+        ------
+        ValueError
+            If the MS file contains data from multiple subarrays.
+        """
+        tb_main = tables.table(filepath, ack=False)
+        time_arr = tb_main.getcol("TIME")
+        # N.b., EXPOSURE is what's needed for noise calculation, but INTERVAL defines
+        # the time period over which the data are collected
```

Comment:
Interesting! Does that mean it supports visibilities that have been integrated together with a gap between them? How does this interact with the observation times?

Reply:
Hypothetically, yes. If that's the case, then the TIME column would actually need to get updated, since it always refers to the temporal midpoint of the visibility, with INTERVAL setting the total length of the window. I believe this gets checked against things that are indexed by time (the one immediate example that pops into mind are the entries in POINTING, but I'm sure there are others).

Reply:
I've got no idea how it'd interact with the other fields in the overall MS structure, but from what's documented, I think it should support that (and in its current state, I don't think it would cause any issues with the data sets that the writer produces).